<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

# Exercise 1
With your knowledge of growing lists, use a for loop with range() and continue to generate a list of the squares of the odd integers from 0 to 9.

In [12]:
for number in range(10):
    if number % 2 == 0: continue        # Removing even numbers
    squared = number ** 2          
    print(squared)

1
9
25
49
81


# Exercise 2
Redo the previous exercise using list comprehension.

In [10]:
[number**2 for number in range(10) if number % 2 != 0]

[1, 9, 25, 49, 81]

# Exercise 3
Use the cell magic command %%timeit to time the previous solutions. Which of the two is faster?

In [18]:
%%timeit
for number in range(10):
    if number % 2 == 0: continue
    squared = number ** 2          

801 ns ± 20.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [19]:
%%timeit
[number**2 for number in range(10) if number % 2 != 0]

884 ns ± 18.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Both codes have similar average execution time of around 800 nanoseconds. However, the first code from exercise 1 is faster, as it has a average execution time of 801 nanoseconds, compared to the second code from exercise 2, with an average execution time of 884 nanoseconds. 

# Exercise 4
The initial quantity of a sample of a radioactive substance is 100 units, and it decays by 5% each year. Use a while loop to determine how long the sample will take to reduce to half its original amount.

In [16]:
sample = 100
year = 0

while sample > 50:              # Half of its original amount: 50 units
    sample *= 0.95
    year += 1
print(f'It takes {year} years for the sample to reduce to half its original amount.')

It takes 14 years for the sample to reduce to half its original amount.


# Exercise 5
The following is data about atmospheric CO levels for several years in a (year, CO2_level) format. The units are ppm. Identify those years that showed an increase of CO of 3 ppm or more compared to the previous year.
Please print out these years along with the corresponding change in concentration.

In [24]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

for i in range(1,len(co2_data)):
    year, level = co2_data[i]
    prev_year, prev_level = co2_data[i-1]
    change = level - prev_level
    if change >= 3:
        print(f'{year} showed an increase of {change:.3f} ppm more than the previous year.')


2016 showed an increase of 3.410 ppm more than the previous year.
